src/Numerics/LinearSolvers/BatchedGeneralizedMinimalResidualSolver.jl

```diff
@@ -0,0 +1,824 @@
+    M, I = @index(Global, NTuple)
+    tmp = zero(eltype(gmres.b))
+    @inbounds for j in 1:i
+        tmp += gmres.Q[M, j, I] * gmres.sol[j, I]
+    end
+    gmres.x[M, I] += tmp # since previously gmres.x held the initial value
+end
+
+# Configuration for Kernels
+"""
+initialize_gmres!(gmres; ndrange = gmres.n, cpu_threads = Threads.nthreads(), gpu_threads = 256)
+
+# Description
+Uses the initialize_gmres_kernel! for initalizing
+
+# Arguments
+- `gmres`: (struct) [OVERWRITTEN]
+
+# Keyword Arguments
+- `ndrange`: (int) or (tuple) thread structure to iterate over
+- `cpu_threads`: (int) number of cpu threads. default = Threads.nthreads()
+- `gpu_threads`: (int) number of gpu threads. default = 256
+
+# Return
+event. A KernelAbstractions object
+"""
+function initialize_gmres!(
+    gmres::BatchedGeneralizedMinimalResidual;
+    ndrange = gmres.n,
+    cpu_threads = Threads.nthreads(),
+    gpu_threads = 256,
+)
+    if isa(gmres.b, Array)
+        kernel! = initialize_gmres_kernel!(CPU(), cpu_threads)
+    else
+        kernel! = initialize_gmres_kernel!(CUDA(), gpu_threads)
+    end
+    event = kernel!(gmres, ndrange = ndrange)
+    return event
+end
+
+"""
+gmres_update!(i, gmres; ndrange = gmres.n, cpu_threads = Threads.nthreads(), gpu_threads = 256)
+
+# Description
+Calls the gmres_update_kernel!
+
+# Arguments
+- `i`: (int) iteration number
+- `gmres`: (struct) gmres struct
+
+# Keyword Arguments
+- `ndrange`: (int) or (tuple) thread structure to iterate over
+- `cpu_threads`: (int) number of cpu threads. default = Threads.nthreads()
+- `gpu_threads`: (int) number of gpu threads. default = 256
+
+# Return
+event. A KernelAbstractions object
+"""
+function gmres_update!(
+    i,
+    gmres;
+    ndrange = gmres.n,
+    cpu_threads = Threads.nthreads(),
+    gpu_threads = 256,
+)
+    if isa(gmres.b, Array)
+        kernel! = gmres_update_kernel!(CPU(), cpu_threads)
+    else
+        kernel! = gmres_update_kernel!(CUDA(), gpu_threads)
+    end
+    event = kernel!(i, gmres, ndrange = ndrange)
+    return event
+end
+
+"""
+construct_solution!(i, gmres; ndrange = size(gmres.x), cpu_threads = Threads.nthreads(), gpu_threads = 256)
+
+# Description
+Calls construct_solution_kernel! for constructing the solution
+
+# Arguments
+- `i`: (int) iteration number
+- `gmres`: (struct) gmres struct
+
+# Keyword Arguments
+- `ndrange`: (int) or (tuple) thread structure to iterate over
+- `cpu_threads`: (int) number of cpu threads. default = Threads.nthreads()
+- `gpu_threads`: (int) number of gpu threads. default = 256
+
+# Return
+event. A KernelAbstractions object
+"""
+function construct_solution!(
+    i,
+    gmres;
+    ndrange = size(gmres.x),
+    cpu_threads = Threads.nthreads(),
+    gpu_threads = 256,
+)
+    if isa(gmres.b, Array)
+        kernel! = construct_solution_kernel!(CPU(), cpu_threads)
+    else
+        kernel! = construct_solution_kernel!(CUDA(), gpu_threads)
+    end
+    event = kernel!(i, gmres, ndrange = ndrange)
```

Comment:
Again, I think you might need a dependencies argument.  You can think about adding a dependencies argument to the `construct_solution!` function.